tecton_client/_internal/data_types.py

```diff
@@ -0,0 +1,48 @@
+from dataclasses import dataclass
+from typing import Dict, List, Optional
+
+
+@dataclass
+class GetFeaturesResult:
+    features: list
+
+
+@dataclass
+class GetFeaturesResponse:
+    result: GetFeaturesResult
+    metadata: Optional[Dict] = None
+
+    @classmethod
+    def from_dict(cls, resp: dict) -> "GetFeaturesResponse":
+        return GetFeaturesResponse(result=GetFeaturesResult(**resp["result"]), metadata=resp.get("metadata"))
```

Comment:
Can we add a `types.MappingProxyType` to each of these to enforce read-only mapping? We'd also update the type annotation from `Dict` to `Mapping`. Ideally we could resolve the type annotation further `Mapping[str, str]`?

Reply:
Not sure I understand this. The pattern here is:
```
resp = await self._client.post(url=self._paths["get_features"], json=request_data)
GetFeaturesResponse.from_dict(resp.json())
```
`resp.json()` is a dict, so the type here needs to be a dict, not a MappingProxyType, unless I misunderstand the suggestion.

Is this maybe meant for `MetadataOptions.to_dict` method?

Reply:
Ah let me give a quick snippet to illustrate.

```
@dataclass
class GetFeaturesResponse:
    result: GetFeaturesResult
    metadata: Optional[Mapping[str, str]] = None

    @classmethod
    def from_dict(cls, resp: dict) -> "GetFeaturesResponse":
        return GetFeaturesResponse(
            result=GetFeaturesResult(**resp["result"]),
            # NOTE: using `types.MappingProxyType` ensures that our stored object is read-only.
            metadata=types.MappingProxyType(resp.get("metadata"))
        )
```

Basically the higher level idea here is that we use `MappingProxyType` to get a dict-like object which is immutable (i.e. we + our consumers can't modify it).